homeassistant/components/greencell/__init__.py

```diff
@@ -0,0 +1,194 @@
+"""Home Assistant integration for Greencell EVSE devices.
+
+This module provides the setup and discovery logic for the Greencell integration:
+- Registers the core integration via async_setup and async_setup_entry.
+- Subscribes to the GREENCELL_DISC_TOPIC for device "hello"/reset announcements.
+- For any newly discovered device (ID not already configured), publishes a QUERY command
+  to prompt the device to send its state and configuration.
+
+Key functions:
+- async_setup(hass, config):
+    Called at Home Assistant startup; installs the discovery listener.
+- async_setup_entry(hass, entry):
+    Called for each config entry; waits for first device message, stores runtime_data,
+    forwards setups to SENSOR platform.
+- setup_discovery_listener(hass):
+    Defines and schedules subscription to GREENCELL_DISC_TOPIC for new devices.
+- wait_for_device_ready(hass, serial, timeout):
+    Subscribes to both DISC_TOPIC and device voltage topic, waits for first message.
+"""
+
+import asyncio
+from collections.abc import Callable
+import json
+from json import JSONDecodeError
+import logging
+from typing import Any
+
+from greencell_client.access import GreencellAccess, GreencellHaAccessLevel
+from greencell_client.elec_data import ElecData3Phase, ElecDataSinglePhase
+
+from homeassistant.components import mqtt
+from homeassistant.components.mqtt import async_subscribe
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady, HomeAssistantError
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    CONF_SERIAL_NUMBER,
+    DISCOVERY_TIMEOUT,
+    DOMAIN,
+    GREENCELL_ACCESS_KEY,
+    GREENCELL_CURRENT_DATA_KEY,
+    GREENCELL_DISC_TOPIC,
+    GREENCELL_POWER_DATA_KEY,
+    GREENCELL_STATE_DATA_KEY,
+    GREENCELL_VOLTAGE_DATA_KEY,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [Platform.SENSOR]
+
+
+def setup_discovery_listener(hass: HomeAssistant) -> Callable[[], None]:
+    """Set up listener for discovery/hello messages; returns unsubscribe function."""
+    unsub: Callable[[], None] | None = None
```

Comment:
What's the point of the discovery happening here? Config entries are tied to a certain device, why do we discover additional ones here?